src/components/modals/PluginApiModal.js

```diff
@@ -1,54 +1,57 @@
-import React, { useEffect } from 'react'
+import React, { useEffect, useContext } from 'react'
 import PropTypes from 'prop-types'
-import uniqueId from 'lodash/uniqueId'
-import { getWebAdminWindow, getWebAdminDocumentBody } from '_/utils/webadmin-dom'
+import { WebAdminModalContext } from '_/utils/react-modals'
 
 import { Modal } from '@patternfly/react-core'
 
 const PluginApiModal = ({
   children,
   title,
   className,
-  modalId = `pluginApiModal-${uniqueId()}`,
+  id,
   isOpen,
   onClose = () => {},
   ...restForModal
 }) => {
+  const modalContext = useContext(WebAdminModalContext)
+
   useEffect(() => {
     if (isOpen) {
-      const targetWindow = getWebAdminWindow()
       const clonedStyles = []
 
-      if (window !== targetWindow) {
+      if (window !== modalContext.targetWindow) {
+        const cssTop = []
+        const cssBottom = []
+
         window.document.querySelectorAll('head style, head link[type="text/css"], head link[rel="stylesheet"]').forEach(style => {
           const cloned = style.cloneNode(true)
-          cloned.setAttribute('data-style-for', modalId)
+          cloned.setAttribute('data-style-for', id)
           clonedStyles.push(cloned)
 
           if (/ui-extensions\/css\/vendor\.[^/]*?css$/.test(cloned.href)) {
-            targetWindow.document.head.insertBefore(cloned, targetWindow.document.head.firstChild)
+            cssTop.push(cloned)
           } else {
-            targetWindow.document.head.appendChild(cloned)
+            cssBottom.push(cloned)
           }
         })
+
+        modalContext.applyCss(cssTop, cssBottom)
       }
 
       return () => {
-        clonedStyles.forEach(style => {
-          targetWindow.document.head.removeChild(style)
-        })
+        modalContext.removeCss(clonedStyles)
       }
     }
-  }, [isOpen, modalId])
+  }, [isOpen, id, modalContext])
```

Comment:
`modalContext`  is created as object literal so it will have different reference with each execution of `createModalContextValue()`. Each reference change would cause re-render and most likely blinking. Fortunately, in our current code it seems to be stable: the root for regular modals is rendered once and modals triggered from dashboard use default context. However,  since this is not obvious, some comment in the code would be helpful.

Reply:
I think it is more obvious in the current iteration.

Reply:
Actually it's still not obvious why `modalContext` is stable. Some comment would clarify that.

Reply:
Done.